packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx

```diff
@@ -130,10 +162,45 @@ export default function useTdpClientCanvas(props: Props) {
   // on the remote machine.
   const onContextMenu = () => false;
 
+  const sendLocalClipboardToRemote = (cli: TdpClient) => {
+    // We must check that the DOM is focused or navigator.clipboard.readText throws an error.
+    // We check that initialTdpConnectionSucceeded so that we don't mistakenly send clipboard data
+    // to a backend that isn't ready for it yet, which would fail silently.
+    if (
+      shareClipboard.current &&
+      document.hasFocus() &&
+      initialTdpConnectionSucceeded.current
+    ) {
+      navigator.clipboard.readText().then(text => {
+        if (text != latestClipboardText.current) {
+          // Wrap in try catch so that lastCopiedClipboardText is only
+          // updated if sendClipboardData succeeds.
+          // eslint-disable-next-line no-useless-catch
+          try {
+            cli.sendClipboardData({
+              data: text,
+            });
+            latestClipboardText.current = text;
+          } catch (e) {
+            throw e;
+          }
+        }
+      });
+    }
+  };
+
+  // Syncs the browser-side's clipboard. See the note about mouseenter in the relevant RFD for why this makes sense:
+  // https://github.com/gravitational/teleport/blob/master/rfd/0049-desktop-clipboard.md#local-copy-remote-paste
+  const onMouseEnter = (cli: TdpClient, e: MouseEvent) => {
+    e.preventDefault();
+    sendLocalClipboardToRemote(cli);
```

Comment:
Do you need to do this on every mouse enter even if you can prove the contents didn't change?

Reply:
We don't; `sendLocalClipboardToRemote` checks if the contents have changed [here](https://github.com/gravitational/webapps/blob/8d3223580614118b3b929f3cba742ac7b9562f79/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx#L175).

That value is also updated when we are sent clipboard data from remote: https://github.com/gravitational/webapps/blob/8d3223580614118b3b929f3cba742ac7b9562f79/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx#L96